server/src/main/java/org/opensearch/action/ActionModule.java

```diff
@@ -954,13 +960,87 @@ protected void configure() {
                 bind(supportAction).asEagerSingleton();
             }
         }
+
+        // register dynamic ActionType -> transportAction Map used by NodeClient
+        bind(DynamicActionRegistry.class).toInstance(dynamicActionRegistry);
     }
 
     public ActionFilters getActionFilters() {
         return actionFilters;
     }
 
+    public DynamicActionRegistry getDynamicActionRegistry() {
+        return dynamicActionRegistry;
+    }
+
     public RestController getRestController() {
         return restController;
     }
+
+    /**
+     * The DynamicActionRegistry maintains a registry mapping {@link ActionType} instances to {@link TransportAction} instances.
+     * <p>
+     * This class is modeled after {@link NamedRegistry} but provides both register and unregister capabilities.
+     *
+     * @opensearch.internal
+     */
+    public static class DynamicActionRegistry {
+        // This is the immutable actions map created during node bootstrap, which
+        // will continue to link ActionType and TransportAction pairs from core and plugin
+        // action handler registration.
+        private Map<ActionType, TransportAction> actions = Collections.emptyMap();
```

Comment:
Please do not use raw type for type with generic type arguments:

```suggestion
        private Map<ActionType<?>, TransportAction<?, ?>> actions = Collections.emptyMap();
```

Same applies for `registry`

Reply:
I can change for `registry` but the raw types are required for the `actions` map.   

The existing code sends to NodeClient with this line:
https://github.com/opensearch-project/OpenSearch/blob/f40fa82e3667825fd0ba8d4bb17b6959ff7cc4f1/server/src/main/java/org/opensearch/node/Node.java#L1115

and the NodeClient version of the map (which is now being wrapped inside the registry) is:
https://github.com/opensearch-project/OpenSearch/blob/f40fa82e3667825fd0ba8d4bb17b6959ff7cc4f1/server/src/main/java/org/opensearch/client/node/NodeClient.java#L60

Reply:
Or I guess I can add the `<?>` to the existing code....

Reply:
Seems to not work:
```
»  org.opensearch.common.inject.ConfigurationException: Guice configuration errors:
»  
»  1) No implementation for java.util.Map<org.opensearch.action.ActionType<?>, org.opensearch.action.support.TransportAction<?, ?>> was bound.
»    while locating java.util.Map<org.opensearch.action.ActionType<?>, org.opensearch.action.support.TransportAction<?, ?>>
```
I think with type erasure we have to use raw types with the injection.

Reply:
:o this is very surprising, will take a look shorlty